core/src/main/java/io/neba/core/resourcemodels/mapping/CustomFieldMappers.java

```diff
@@ -0,0 +1,114 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+package io.neba.core.resourcemodels.mapping;
+
+import io.neba.api.resourcemodels.AnnotatedFieldMapper;
+import io.neba.core.resourcemodels.metadata.MappedFieldMetaData;
+import io.neba.core.util.ConcurrentDistinctMultiValueMap;
+import org.springframework.stereotype.Service;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import static java.util.Collections.emptyList;
+
+/**
+ * @author Olaf Otto
+ */
+@Service
+public class CustomFieldMappers {
+    /**
+     * @author Olaf Otto
+     */
+    public static class AnnotationMapping {
+        private final Annotation annotation;
+        private final AnnotatedFieldMapper mapper;
+
+        public AnnotationMapping(Annotation annotation, AnnotatedFieldMapper mapper) {
+            this.annotation = annotation;
+            this.mapper = mapper;
+        }
+
+        public Annotation getAnnotation() {
+            return annotation;
+        }
+
+        public AnnotatedFieldMapper getMapper() {
+            return mapper;
+        }
+    }
+
+    private static final Collection<AnnotationMapping> NULL = emptyList();
+    private final ConcurrentDistinctMultiValueMap<Field, AnnotationMapping> cache
+          = new ConcurrentDistinctMultiValueMap<Field, AnnotationMapping>();
+    private final ConcurrentDistinctMultiValueMap<Class<? extends Annotation>, AnnotatedFieldMapper> fieldMappers
+          = new ConcurrentDistinctMultiValueMap<Class<? extends Annotation>, AnnotatedFieldMapper>();
+
+    public synchronized void add(AnnotatedFieldMapper<?, ?> mapper) {
+        if (mapper == null) {
+            throw new IllegalArgumentException("Method argument mapper must not be null.");
+        }
+        this.fieldMappers.put(mapper.getAnnotationType(), mapper);
+        this.cache.clear();
+    }
+
+    public synchronized void remove(AnnotatedFieldMapper<?, ?> mapper) {
+        if (mapper == null) {
+            throw new IllegalArgumentException("Method argument mapper must not be null.");
+        }
+        this.fieldMappers.remove(mapper.getAnnotationType());
+        this.cache.clear();
+    }
+
+    public Collection<AnnotationMapping> get(MappedFieldMetaData metaData) {
+        if (metaData == null) {
+            throw new IllegalArgumentException("Method argument metaData must not be null.");
+        }
+
+        Collection<AnnotationMapping> mappers = this.cache.get(metaData.getField());
+
+        if (mappers == NULL) {
+            return NULL;
```

Comment:
This if block is not necessary, if mappers == NULL, it is not null, so it was returned within the next if block anyway. Btw, NULL is quite confusing for a Java programmer, I think. Why not EMPTY_LIST or so?